Making preimage propagation more reliable

ferencdg opened the pull request:
fixes: #2217
fixes: #2246

Commit:
Added preimage sharing during SCP protocol run

Validators should proactively exchange preimage information, if they
detect that other validators might have the preimages of validators that our
current validator doesn't have.

fixes: #2217

ferencdg commented on source/agora/api/Validator.d:
```diff
@@ -120,4 +121,22 @@ public interface API : agora.api.FullNode.API
+            GET /preimages_for_enroll_keys
+
+    ***************************************************************************/
+
```
we could be more efficient here by only sharing the indices, but this will unlikely to cause any real problems

ferencdg commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -873,13 +873,6 @@ extern(D):
             return ValidationLevel.kInvalidValue;
         }
 
```
there were 2 problems with this code:
1. we check for the same thing in the ValidateConsensusData too
2. making the check here would have caused us not to record the missing validators, so I couldn't have implemented my ticket

Because of 2., it is not in a separate commit

ferencdg commented on source/agora/test/SlashingMisbehavingValidator.d:
```diff
@@ -115,23 +115,6 @@ unittest
 ///     some failures.
 unittest
 {
-    static class LazyAPIManager : TestAPIManager
```
needed to reuse this in my ticket, so that is why there is no separate commit

Geod24 replied:
~So this change can be reverted nowadays (or put in a separate commit, if you wish).~ Ignore this, I misread the diff.
By the way, it would make sense to have the move as a separate commit, explaining the rationale, which precedes the commit that uses it. It does make reviewing much easier.

ferencdg commented on source/agora/node/Ledger.d:
```diff
@@ -722,18 +722,19 @@ public class Ledger
     ***************************************************************************/
 
-    public bool isSelfSlashing (in Height height, in ConsensusData data) @safe nothrow
+    public string isSelfSlashing (in Height height, in ConsensusData data) @safe nothrow
```
see https://github.com/bosagora/agora/pull/2244/files#r664243457 for why it is not in a separate commit

hewison-chris commented on source/agora/node/Ledger.d:
```diff
@@ -1166,6 +1166,10 @@ public class Ledger
+    public Set!Hash getEnrollKeysForUnknownPreimages () @safe nothrow
+    {
+        return this.enrolls_keys_for_unknown_preimages;
+    }
```
Please update the commit message to match the actual changes.

ferencdg replied:
I staged the wrong chunk there :( fixed by now

hewison-chris replied:
But shouldn't the new function be in the next commit? In fact I don't really see why this adding `nothrow` is not done in the commit that requires it.

ferencdg replied:
right, squashed it

hewison-chris commented on source/agora/network/Client.d:
```diff
@@ -524,6 +524,27 @@ public class NetworkClient
 
+    /***************************************************************************
+
+        Returns he preimages for the specified enroll keys for the specified height.
```
```suggestion
        Returns the preimages for the specified enroll keys and height.
```

hewison-chris commented on source/agora/network/Manager.d:
```diff
@@ -976,6 +976,36 @@ public class NetworkManager
+
+        Params:
+            ledger = ledger instance
+            height = height
```
`height` is not a param.

hewison-chris commented on source/agora/network/Manager.d:
```diff
@@ -976,6 +976,36 @@ public class NetworkManager
+
+    ***************************************************************************/
+
+    public void retrievePreimages (Ledger ledger) @safe nothrow
```
```suggestion
    public void retrievePreimages (in Ledger ledger) @safe nothrow
```

ferencdg replied:
sadly 'in' makes the ledger constant, but I need to call mutable variables methods like: 
```D
ledger.enrollment_manager.addPreimage(preimage_info)
```

hewison-chris commented on source/agora/network/Manager.d:
```diff
@@ -976,6 +976,36 @@ public class NetworkManager
+
+    public void retrievePreimages (Ledger ledger) @safe nothrow
+    {
+
```
```suggestion
```

hewison-chris replied:
This extra line is still there.

hewison-chris commented on source/agora/network/Manager.d:
```diff
@@ -976,6 +976,36 @@ public class NetworkManager
+
+        foreach (peer; this.peers[])
+        {
+            if (enroll_keys.empty())
```
Why don't we check this before the `foreach`?

ferencdg replied:
enroll_keys is also changed inside the foor loop, so we need to check it there too

I could check it in front of the loop as well, but entering the foreach have almost no performance impact and for me, the code looks cleaner this way.

hewison-chris reviewed (changes requested):
LGTM, just a few small changes requested.

linked0 reviewed (commented):
Overall LGTM but first commit has a member variable that is introduced in the next commit, which is `enrolls_keys_for_unknown_preimages`.

ferencdg commented:
fixed everything else that was suggested

hewison-chris reviewed (approved):
LGTM

kchulj commented on source/agora/node/Ledger.d:
```diff
@@ -722,18 +722,19 @@ public class Ledger
         a node itself
 
         Params:
+            height = the heigh for which we want to check self slashing
```
```suggestion
            height = the height for which we want to check self slashing
```

Geod24 commented on dub.json:
```diff
@@ -36,7 +36,6 @@
 	"gdc": "no",
 	"ldc": "~>1.26"
     },
-
```
Unrelated

Geod24 commented on source/agora/api/Validator.d:
```diff
@@ -120,4 +121,22 @@ public interface API : agora.api.FullNode.API
+
+    ***************************************************************************/
+
+    public PreImageInfo[] getPreimagesForEnrollKeys (Set!Hash enroll_keys, ulong height);
```
Hum, how does a set get serialized ? I'd expect a `Hash[]` here.

Geod24 replied:
Also, `FullNode` already has `getPreimages`. This is equivalent to `getPreimages(height, height)`, isn't it ?

ferencdg replied:
it is not exactly equivalent, because we don't expect many missing preimages, only a few, but probably none at all. getPreimages(height, height) returns every preimage

if and attacker would claim that he knows the missing preimage for NodeA (where in fact NodeA's preimage is missing) getPreimages(height, height) would return all preimages from all nodes

Geod24 replied:
But is it worth the extra endpoint ?

ferencdg replied:
> Hum, how does a set get serialized ? I'd expect a `Hash[]` here.

it is a hash array I think
```D
    public string toString () const
    {
        import std.format;

        return format("%(%s, %)",
                      this._set.byKeyValue.filter!(tup => tup.value)
                      .map!(tup => tup.key));
    }
```

ferencdg replied:
> But is it worth the extra endpoint ?

adding an extra endpoint is not ideal (as it is an extra burden on a new client which is not implemented in DLang), but I think it is worth it in this case

if we really want to remove a useless endpoint then I think it should be 

public PreImageInfo getPreimage (Hash enroll_key);

1. this endpoint requires clients to query for each enroll keys
2. height is assumed to be the highest which is not reliable

Geod24 commented on source/agora/test/PreimageSharing.d:
```diff
@@ -0,0 +1,63 @@
+module agora.test.PreimageSharing;
+
+/*******************************************************************************
+
+    Contains tests for sharing preimages during SCP protocol run.
+
+    Copyright:
+        Copyright (c) 2019-2021 BOSAGORA Foundation
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
```
```suggestion
/*******************************************************************************

    Contains tests for sharing preimages during SCP protocol run.

    Copyright:
        Copyright (c) 2019-2021 BOSAGORA Foundation
        All rights reserved.

    License:
        MIT License. See LICENSE for details.

*******************************************************************************/

module agora.test.PreimageSharing;
```

Geod24 reviewed (changes requested):
The API endpoint is duplicated. The rest looks good, will do a second pass once duplication is removed.

ferencdg commented:
rebase #1...

ferencdg commented:
@Geod24 fixed everything, I think instead of not adding the new getPreimagesForEnrollKeys() method, we should remove the dumber version of it called getPreimage(), which I could do in a separate PR

Geod24 commented:
Well nothing prevents a node from returning more preimages than requested, so the API is only useful for the server, if the client is well behaving.
Having 2 (or 3) endpoints for the same thing is indeed quite ugly. What we currently have:
- `getPreimage` => Returns a single UTXO's preimage, at a given height;
- `getPreimages(from, to)` => Return all preimages for validator between `from` and `to`;
- `getPreimagesForEnrollKeys` => Return all peimages for a certain set of key;

Now there's a few things to take into account:
- A node only needs to return a single `PreImageInfo` for a given UTXO;
- That used not to be the case, which motivated `getPreimages(from, to)`;
- I *think* nodes should return the latest known pre-image regardless of the requested height, and the client should adjust;
- As a result, providing the height to any of those endpoints is *pointless* unless we don't know the validator set;
- If there isn't a case in which we don't know the validator set, we could have an endpoint `getPreimages(Hash[] utxos)` and call it a day;
- But that would require sending *all* validators to a node, which would lead to pretty large (and useless) queries;
- What we could do instead is to have `getPreimages(Hash[] only = null);` where it returns either all preimages (for the node's current validator set) or the `only` subset requested - assumed to be small;

Note that this *could* create issues when catching up, as catching up requires preimages. However, a trivial fix is to request any new validator immediately: Since they tend to be long lived, it will only leads to sporadic, single-argument queries.

Thoughts ?

ferencdg commented:
getPreimagesForEnrollKeys => Return all peimages for a certain set of key; - this is not true, it only returns for a specified height

ferencdg commented:
'Well nothing prevents a node from returning more preimages than requested, so the API is only useful for the server, if the client is well behaving.'

but the solution to return ALL preimages is not useful even for well-behaving clients. I think what you said is true for any endpoint, a non well-behaving node can return an endless stream of garbage as long as it can be deserialized.

Geod24 commented:
> this is not true, it only returns for a specified height

Why would we not return the latest preimage ?

ferencdg commented:
> > this is not true, it only returns for a specified height
> 
> Why would we not return the latest preimage ?

this function is called during the short run of the nomination protocoll, we cannot just ask for the latest one as the latest one is ususally 5 cycles ahead

Geod24 commented:
> this function is called during the short run of the nomination protocoll, we cannot just ask for the latest one as the latest one is ususally 5 cycles ahead

If you know the one 5 cycles ahead, you know the current one and you don't have to request it again.

ferencdg commented:
> > this is not true, it only returns for a specified height
> 
> Why would we not return the latest preimage ?

I also think that if the requesting node has to depend on the current context of the REQUESTED node (what is the maximum known preimage received by the REQUESTED node so far), then it is much harder to reason about code correctness in the requesting node. For example what if the requested node only knows the preimage for height 5, but the requested node has the preimage for height 7. Now the requested node will reject the returned preimage.

Geod24 commented:
> I also think that if the requesting node has to depend on the current context of the REQUESTED node (what is the maximum known preimage received by the REQUESTED node so far), then it is much harder to reason about code correctness in the requesting node. For example what if the requested node only knows the preimage for height 5, but the requested node has the preimage for height 7. Now the requested node will reject the returned preimage.

Discussed IRL, and cleared up I hope.

Commit:
Removed FullNode.getPreimage

FullNode.getPreimages was introduced previously and made
FullNode.getPreimage obsolate

ferencdg commented:
made the requested changes

Geod24 commented on source/agora/api/FullNode.d:
```diff
@@ -260,6 +260,25 @@ public interface API
+
+    ***************************************************************************/
+
+    public PreImageInfo[] getPreimagesForEnrollKeys (Set!Hash enroll_keys);
```
Set it to `null` by default, and rename to `getPreimages`

Geod24 replied:
Also I believe we could just use `Hash[]` here instead of a `Set` ?

ferencdg replied:
1. there are a lot of convenient function on set like setDifference that I use in many places to make the code more clear
2. followed this: getTransactions (Set!Hash tx_hashes); which I think the correct approach

Geod24 replied:
`setDifference` works on range, so it will work on arrays. Also, it assumes the input is sorted, which I don't think is the case here ?

Geod24 replied:
Regarding point 2, I don't *mind* having `Set` in the API. But:
- [x] It needs to be convenient to use from other languages: Check, because it serializes just like an array;
- [ ] It needs to be convenient to call from our own code: This is the current problem, because calling this with a single value is 3-4 lines of code;
- [x] It needs not to incur more penalty (speed, memory usage, etc...) than an array: I believe this is more-or-less the case right now (not exactly, because one can't manage the underlying memory of an AA, but it's a price we decided to pay), but it's something to keep in mind, especially if you solve the second point;

ferencdg replied:
'rename to getPreimages' - isn't that a bit ambiguous for a REST interface as we already have 

public override PreImageInfo[] getPreimages (ulong start_height,
        ulong end_height)
        
can Vibe.d even handle this (2 REST interfaces with the same endpoint name)?

omerfirmak replied:
Vibe.d serializes Sets as dictionaries btw.

ferencdg replied:
>  Also, it assumes the input is sorted, which I don't think is the case here ?

'it assumes the input is sorted' - it is sorted in our current implementation (checked by a quick look)

`It needs to be convenient to call from our own code: This is the current problem, because calling this with a single value is 3-4 lines of code;` 

extra memory allocation?

Geod24 commented on source/agora/api/FullNode.d:
```diff
@@ -260,6 +260,25 @@ public interface API
+            If the requested node doesn't know the preimage for a specific
+            enroll key, then it will not be included in the result.
+            If the set of enroll_keys is empty, then all preimages known to the
+            node are returned.
```
Should be part of the param documentation

Geod24 commented on source/agora/api/FullNode.d:
```diff
@@ -260,6 +260,25 @@ public interface API
+        Returns:
+            The preimages for the specified enroll keys.
+            If the requested node doesn't know the preimage for a specific
+            enroll key, then it will not be included in the result.
```
This looks like some of this should be the title / description, not in the `Returns` section.

Geod24 commented on source/agora/api/Validator.d:
```diff
@@ -47,6 +47,7 @@ import agora.consensus.data.ValidatorBlockSig;
 import agora.crypto.Hash;
 import agora.crypto.Key;
 import agora.crypto.Schnorr;
+import agora.common.Set;
```
Remove

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -1260,6 +1275,11 @@ public class Ledger
                 missing_validators_lower_bound ~= cast(uint)idx;
         }
 
+        // trying to retrieve preimages that - based on the consensus data -
+        // was shared with other nodes
+        foreach (const validator_ind; setDifference(missing_validators_lower_bound, missing_validators))
+            this.enrolls_keys_for_unknown_preimages.put(keys[validator_ind]);
+
```
This method was not originally modifying any class member, which means it was idempotent.
However, with this addition, it no longer is, and there's an assumption that it will only be called once per (I'm not sure what).
Also, I couldn't find the place where this was reset.

ferencdg replied:
reset is in NetworkManager.retrievePreimages()
```D
        () @trusted {enroll_keys.clear();}();
```

ferencdg replied:
'However, with this addition, it no longer is, and there's an assumption that it will only be called once per' 

it is a Set!Hash - there is no assumption (one of the reason why I used Sets)

ferencdg replied:
'This method was not originally modifying any class member, which means it was idempotent.'

suggestion where it should be?

Geod24 replied:
Discussed offline, we're going with this approach.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -1173,6 +1174,20 @@ public class Ledger
+    ***************************************************************************/
+
+
+    public Set!Hash getEnrollKeysForUnknownPreimages () @safe nothrow
```
Instead of this member function, I think you can just call `getValidators` (since you have access to the Ledger) and check the PI height.

ferencdg replied:
in enrolls_keys_for_unknown_preimages we store what other validators reported regarding for what validators they have preimage for(and we don't)

I don't think our validator's ledger has that information.

Geod24 commented on source/agora/test/PreimageSharing.d:
```diff
@@ -0,0 +1,63 @@
+import core.stdc.stdint;
+import core.thread;
+
+import geod24.Registry;
```
There's a few imports there that are superfluous

Geod24 commented on source/agora/test/PreimageSharing.d:
```diff
@@ -0,0 +1,63 @@
+    // block should be created even though the validators didn't reveal their
+    // preimages in advance, as preimage sharing is also done during SCP run
+    network.expectHeight(Height(1));
+
```
Trailing line

Geod24 commented on source/agora/test/PreimageSharing.d:
```diff
@@ -0,0 +1,63 @@
+    };
+
+    // set up nodes
+    auto network = makeTestNetwork!LazyAPIManager(conf);
```
Looking at this test, it shouldn't work. `NoPreImageVN` was designed to *not* reveal pre-images, unless told to do so. That's why `getPreimage was overriden.

ferencdg replied:
You can see in the comment why it works
```D
/// Please note that preimages can still be shared during nomination, if the
/// TestConf.preimage_catchup_interval is set to a low value.
public class NoPreImageVN : TestValidatorNode
```

Geod24 replied:
But then, the other test that uses this class, which relies on preimages *really* not being propagated has its assumptions gutted, doesn't it ?

ferencdg replied:
the default value for TestConf.preimage_catchup_interval is set to the value:

public Duration preimage_catchup_interval = 100.seconds; 

so none of the existing test fails. We could be set it to much higher.

preimage_catchup_interval is more like a switch in NoPreimageVN, if you want no preimage at all, set it to a high value(leave it at default), but you have to option to customize it and share preimages during nomination.

ferencdg replied:
> the default value for TestConf.preimage_catchup_interval is set to the value:
> 
> public Duration preimage_catchup_interval = 100.seconds;
> 
> so none of the existing test fails. We could be set it to much higher.
> 
> preimage_catchup_interval is more like a switch in NoPreimageVN, if you want no preimage at all, set it to a high value(leave it at default), but you have to option to customize it and share ceratain preimages during nomination.

Geod24 replied:
Not sure I follow. `NoPreImageVN` should *never* return pre-images if the boolean is set to `false`.
Obviously the rest of the code is unchanged, so it tries to catch-up at a regular interval, but since all validators in the network are `NoPreImageVN`, I would not expect any pre-image to be revealed.

Geod24 replied:
> so none of the existing test fails.

They don't fail, but they might start to succeed for the wrong reason.

ferencdg replied:
> Not sure I follow. `NoPreImageVN` should _never_ return pre-images if the boolean is set to `false`.
> Obviously the rest of the code is unchanged, so it tries to catch-up at a regular interval, but since all validators in the network are `NoPreImageVN`, I would not expect any pre-image to be revealed.

NoPreImageVN - there are(or will be) at least 3 ways to reveal preimages
1. sharing preimages in advance in a push way
2. sharing preimages during nomination in a pull way
3. preimage catchup

instead of having NoPreImageVN we should allow all those 3 methods to be enabled/disabled (and not just by a bool value), having 

preimage_catchup_interval = 100.seconds;

should be the first step for that

I don't think we want 2<sup>3</sup> different classes for all the combinations

ferencdg replied:
> > so none of the existing test fails.
> 
> They don't fail, but they might start to succeed for the wrong reason.

:) I don't get this, if we set 
preimage_catchup_interval = 1000000.seconds;

then the only way for a test to succeed for the wrong reason is that the test actually ran 1000000 seconds and succeeded because of nomination preimage sharing

ferencdg replied:
if you have any suggestion how you would like to see this, please share in detail and I will implement it that way

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -710,7 +710,14 @@ public class TestAPIManager
+                    retryFor(
+                        ()
+                        {
+                            Set!Hash enroll_keys;
+                            enroll_keys.put(enroll.utxo_key);
+                            auto preimages = this.clients[idx].getPreimagesForEnrollKeys(enroll_keys);
```
You could just use `[enroll.utxo_key]` if it was a `Hash`.

Geod24 replied:
`Set!Hash.from(enroll.utxo_key.only)` should do it ([`only`](https://dlang.org/phobos/std_range.html#only) is in `std.range`).

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -1916,6 +1923,10 @@ public struct TestConf
     /// Transaction put into the relay queue will expire, and will be removed
     /// after `relay_tx_cache_exp`.
     Duration relay_tx_cache_exp = 60.minutes;
+
+    /// How often the validator should try to catchup for the preimages for the
+    /// next block
+    public Duration preimage_catchup_interval = 100.seconds;
```
You inspired https://github.com/bosagora/agora/pull/2276 BTW :)

ferencdg commented:
@Geod24 pls let me know if you disagree with the comments, so I can make the change

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -2201,16 +2233,6 @@ public class NoPreImageVN : TestValidatorNode
             super.onPreImageRevealTimer();
     }
 
-    public override PreImageInfo getPreimage (Hash enroll_key)
-    {
-        if (enroll_key != this.enroll_man.getEnrollmentKey())
-            return super.getPreimage(enroll_key);
-        else if (atomicLoad(*this.reveal_preimage))
-            return super.getPreimage(enroll_key);
-        else
-            throw new Exception("No such PreImage");
-    }
-
```
Yeah, that should not be removed. It's the whole point of the class.

ferencdg replied:
> Yeah, that should not be removed. It's the whole point of the class.
👀 
 
getPreimage (Hash enroll_key) 
was removed from the FullNode's interface...

However it is true that I should have added the new method called
getPreimageWhatever (Set!Hash enroll_keys)

ferencdg commented:
rebase #2 and fixed according to offline discussion

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -1941,8 +1941,13 @@ public struct TestConf
 
         // Always set to true, cannot be overriden, but also set here for clarity
         testing: true,
+
```
Remove

Geod24 reviewed (approved):
Just fix the tiny style issue and feel free to merge.
